File: LeetCode/Java/557.reverse-words-in-a-string-iii.java
/*
 * @lc app=leetcode id=344 lang=java
 *
 * [344] Reverse String
 */
class Solution {
    public void reverse(StringBuilder s, int i, int j) {
        while (i < j) {
            char ch = s.charAt(i);
            s.setCharAt(i++, s.charAt(j));
            s.setCharAt(j--, ch);
        }
    }

    public String reverseWords(String s) {
        StringBuilder str = new StringBuilder(s);

        int i = 0;
        for (int j = 0; j < str.length(); j++) {
            if (str.charAt(j) == ' ') {
                reverse(str, i, j - 1);
                i = j + 1;
            }
        }
        reverse(str, i, str.length() - 1);
        return str.toString();
    }
}

File: LeetCode/CPP/557.reverse-words-in-a-string-iii.cpp
#include <bits/stdc++.h>
using namespace std;
/*
 * @lc app=leetcode id=557 lang=cpp
 *
 * [557] Reverse Words in a String III
 */

// @lc code=start
class Solution
{
public:
    void reverse(string &s, int i, int j)
    {
        while (i < j)
        {
            swap(s[i++], s[j--]);
        }
    }
    string reverseWords(string s)
    {
        int i = 0;
        for (int j = 0; j < s.size(); j++)
        {
            if (isspace(s[j]))
            {
                reverse(s, i, j - 1);
                i = j + 1;
            }
        }
        reverse(s, i, s.size() - 1);
        return s;
    }
};
// @lc code=end
